ament_clang_tidy/ament_clang_tidy/main.py

```diff
@@ -0,0 +1,290 @@
+        return 1
+
+    # invoke clang_tidy
+    with open(args.config_file, 'r') as h:
+        content = h.read()
+    data = yaml.safe_load(content)
+    style = yaml.dump(data, default_flow_style=True, width=float('inf'))
+    cmd = [clang_tidy_bin,
+           '--config=%s' % style]
+    if args.explain_config:
+        cmd.append('--explain-config')
+    if args.fix_errors:
+        cmd.append('--fix-errors')
+    if args.export_fixes:
+        cmd.append('--export-fixes')
+        cmd.append(args.export_fixes)
+    cmd.extend(files)
+    cmd.append('--')
+    try:
+        output = subprocess.check_output(cmd).strip().decode()
+    except subprocess.CalledProcessError as e:
+        print("The invocation of '%s' failed with error code %d: %s" %
+              (os.path.basename(clang_tidy_bin), e.returncode, e),
+              file=sys.stderr)
+        return 1
+
+    # output errors
+    report = {}
+    complete_filenames = []
+
+    for filename in files:
+        report[filename] = []
+        complete_filename = os.path.join(
+            os.getcwd(), filename)
+        complete_filenames.append(complete_filename)
+
+    error_re = re.compile(r'\[.*?\]')
+    file_pairs = dict(zip(complete_filenames, files))
+
+    current_file = None
+    new_file = None
+    data = {}
+
+    for line in output.splitlines():
+        # error found
+        if line[0] == '/' and error_re.search(line):
+            for filename in complete_filenames:
+                if filename in line:
+                    new_file = file_pairs[filename]
+                    if current_file is not None:
+                        report[current_file].append(copy.deepcopy(data))
+                        data.clear()
+                    current_file = new_file
+            error_msg = find_error_message(line)
+            line_num, col_num = find_line_and_col_num(line)
+            data['line_no'] = line_num
+            data['offset_in_line'] = col_num
+            data['error_msg'] = error_msg
+        else:
+            data['code_correct_rec'] = data.get('code_correct_rec', '') + line + '\n'
+
+    report[current_file].append(copy.deepcopy(data))
+
+    if args.xunit_file:
+        folder_name = os.path.basename(os.path.dirname(args.xunit_file))
+        file_name = os.path.basename(args.xunit_file)
+        suffix = '.xml'
+        if file_name.endswith(suffix):
+            file_name = file_name[0:-len(suffix)]
+            suffix = '.xunit'
+            if file_name.endswith(suffix):
+                file_name = file_name[0:-len(suffix)]
+        testname = '%s.%s' % (folder_name, file_name)
+        xml = get_xunit_content(report, testname, time.time() - start_time)
+        path = os.path.dirname(os.path.abspath(args.xunit_file))
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(args.xunit_file, 'w') as f:
+            f.write(xml)
+    return
+
+
+def find_executable(file_names):
```

Comment:
Hm.  The duplication of `find_executables`, `get_files`, `get_xunit_content`, etc. with other ament linters is pretty unfortunate.  However, I see that is how we have done it for all of the linters, and I think it deserves a different patch.  I'll request that you open an issue to investigate how to share more code between them.

Reply:
I was thinking the same thing implementing this section - I'll open the issue shortly

Reply:
https://github.com/ament/ament_lint/issues/157 is the new issue.